server/partitions.go

```diff
@@ -0,0 +1,343 @@
+				}
+			}
+		}
+	}
+}
+
+// We use a channel subscription in order to minimize the number
+// of go routines for all the explicit pub subscriptions.
+// This go routine simply pulls a message from the channel and
+// invokes processClientPublish(). We may have slow consumer issues,
+// if so, will have to figure out another way.
+func (p *partitions) postClientPublishIncomingMsgs() {
+	defer p.s.wg.Done()
+	for {
+		select {
+		case <-p.quitCh:
+			return
+		case m := <-p.msgsCh:
+			p.s.processClientPublish(m)
+		}
+	}
+}
+
+// Create the internal subscriptions on the list of channels.
+func (p *partitions) initSubscriptions() error {
+	// NOTE: Use the server's nc connection here, not the partitions' one.
+	for _, channelName := range p.channels {
+		pubSubject := fmt.Sprintf("%s.%s", p.s.info.Publish, channelName)
+		if _, err := p.s.nc.ChanSubscribe(pubSubject, p.msgsCh); err != nil {
+			return fmt.Errorf("could not subscribe to publish subject %q, %v", channelName, err)
+		}
+	}
+	// Add a dedicated subject for when we try to schedule a control
+	// message to be processed after all messages from a given client
+	// have been processed.
+	p.ctrlMsgSubject = fmt.Sprintf("%s.%s", p.s.info.Publish, nats.NewInbox())
+	if _, err := p.s.nc.ChanSubscribe(p.ctrlMsgSubject, p.msgsCh); err != nil {
+		return fmt.Errorf("could not subscribe to subject %q, %v", p.ctrlMsgSubject, err)
+	}
+	return nil
+}
+
+// Sends a request to the rest of the cluster and wait a bit for
+// responses (we don't know if or how many servers there may be).
+// No server lock used since this is called inside RunServerWithOpts().
+func (p *partitions) checkChannelsUniqueInCluster() error {
+	// We will use a subscription on an inbox to get the replies
+	replyInbox := nats.NewInbox()
+	replySub, err := p.nc.SubscribeSync(replyInbox)
+	if err != nil {
+		return fmt.Errorf("unable to create sync subscription: %v", err)
+	}
+	defer replySub.Unsubscribe()
+	// Send our list
+	if err := p.sendChannelsList(replyInbox); err != nil {
+		return fmt.Errorf("unable to send channels list: %v", err)
+	}
+	// Since we don't know how many servers are out there, keep
+	// calling NextMsg until we get a timeout
+	for {
+		reply, err := replySub.NextMsg(partitionsNextMsgTimeout)
+		if err == nats.ErrTimeout {
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("unable to get partitioning reply: %v", err)
+		}
+		resp := spb.CtrlMsg{}
+		if err := resp.Unmarshal(reply.Data); err != nil {
+			return fmt.Errorf("unable to decode partitioning response: %v", err)
+		}
+		if len(resp.Data) > 0 {
+			return fmt.Errorf("channel %q causes conflict with channels on server %q",
+				string(resp.Data), resp.ServerID)
+		}
+	}
+}
+
+// Sends the list of channels to a known subject, possibly splitting the list
+// in several requests if it cannot fit in a single message.
+func (p *partitions) sendChannelsList(replyInbox string) error {
+	sendReq := func(channels []string) error {
```

Comment:
No because in the temp buffer done outside of this function, there is one channel too many, which brought the size above the threshold. This is why I used a temporary buffer in which I encode channels (adding 1 at a time) until the overall size is greater than the limit. I therefore *need* to re-encode with the known number of channels that fit.
Regardless, I am revisiting this whole function. Will resubmit later.